### ISSUE TITLE
output from fluor_fvfm

### ISSUE THREAD

**Issue by dschneiderch** [danforthcenter/plantcv#407]
Is there a reason that `fluor_fvfm` outputs fv instead of fvfm? also, I think it would make more sense to call this function `fluor_YII()` or even `psII_YII` since you can compute YII for various conditions. Would it be worth a subpackage - it's possible that future use cases could call for qI, qE, etc that are specific to PSII measurements?
currently  i was going to contribute a NPQ function in which case I would presumably output `Fm/Fm' - 1` so i wanted to check with you about the general direction of these functions.

**Comment by nfahlgren** [comment]
Nope, we should output the Fv/Fm image rather than Fv. 

I have also been thinking about renaming the function, but I was thinking it should start with `analyze_` to be consistent with (most) of the other functions that output measurements. In the documentation we refer to the data as PSII related, the references to FLU or FLUOR are because that's the name of the camera system in the LemnaTec software. Since there are other types of FLU camera systems that have other purposes it makes sense to make it clearer what kind of data this function uses.

A subpackage could make sense if there are multiple functions with different inputs. Looks like NPQ would need Fm and Fm' so that makes sense as a separate function.

What if we named the functions given the terminology laid out by Baker (2008), Table 1 specifically [1]? I think while Fv/Fm is used to estimate YII maybe it's clearer to name the function after how we measure YII rather than as YII directly?

The only other consideration is that in the Python community function names should be lowercase, though it's only a convention.

1. Baker NR. 2008. Chlorophyll fluorescence: A probe of photosynthesis in vivo. Annual review of plant biology 59:89–113. DOI: 10.1146/annurev.arplant.59.032607.092759.

**Comment by dschneiderch** [comment]
Ok, `fluor_fvfm.py` can be changed easily enough.
I think following Baker is a good idea (also [Murchie and Lawson 2013](https://academic.oup.com/jxb/article/64/13/3983/436509)  is a nice followup with more practical advice). However, I’m not sure how function naming should work. We don’t need separate functions for Fv/Fm, Fv’/Fm’, Fq’/Fm’. Internally I refer to them as fvfm, fvpfmp, etc but that seems clunky. What about `analyze_psiiefficiency()`?

We could use:
```
analyze_psiiefficiency()   - for psiII and friends
analyze_opencenters()   - qL
analyze_quenching()      - qP
analyze_npq()                 - npq
```
We also need Fo' but I'm not sure what to call that. 

Ultimately, I think it might be more intuitive to have
```
analyze_yii()
analyze_ql()
analyze_qp()
analyze_npq()
analyze_fop()   
```
That last one needs some work.

For `analyze_yii()` I would propose we include an argument for the parameter so it can be saved properly. so for fvfm you might call `analyze_yii(fdark, fmin, fmax, parameter = 'Fv/Fm')`   for fvfm  or `analyze_yii(fdark,fmin,fmax,parameter = 'Fv'/Fm'') or `analyze_yii(fdark,fmin, fmax, parameter = 'YII')` where `parameter` can be whatever the user wants to save.

**Comment by nfahlgren** [comment]
This sounds pretty good. 

I see the advantage of having `analyze_yii` with the parameter input so users can label the output as they choose since both Fv/Fm and Fv'/Fm' have the same inputs (other then the dark/light adapted nature of the plants). I was thinking the label would be problematic because we would have to normalize it but I guess we don't really use it in a filename, just the histogram figure labels. The debug image could be named with something like the suffix `_YII.png`. Otherwise, when saving the histogram the user can call it anything they like.

I think like you suggested in the initial post that a subpackage would help organize these functions. Would it make sense to call it `psii`? That would give us module paths like:

```
plantcv.psii.analyze_yii()
plantcv.psii.analyze_ql()
plantcv.psii.analyze_qp()
plantcv.psii.analyze_npq()
plantcv.psii.analyze_fop()
```

Another consideration is whether `fdark` is needed here. It sounded like the WSU system doesn't create an Fdark image right? We only use it for Q/C and Q/C could be a separate function if it would be convenient to drop the input on the analyze function(s).

**Comment by dschneiderch** [comment]
I haven’t worked much with your new output format yet. Would it be possible to save mean fv/fm and std fvfm in output in addition to mean fv’/fm’ and std fv’/fm’? Part of my rationale for the parameter input would be to able to separate values from different calls to analyze_yii()? I often work with induction curves so I usually output multiple mean values of yii for different dark/light adapted states. I label them FvFm, Y1_Alon, Y2_Alon, or FvFm, t40, t60, t80 etc.

Yeah, I thought a subpackage would help keep them organized. Is there any phenomics analyzing PSI, or other aspect of photosynthesis that doesn’t fit under psii? Wondering if we should generalize with `plantcv.photosynthesis.analyze_yii()` or otherwise.

Correct we don’t output fdark from our instrument so I pass np.zeroes_like(img) as the first argument. With or without fdark is fine by me.

**Comment by nfahlgren** [comment]
`plantcv.photosynthesis` works for me.

One of the advantages of the new output format is it makes it way easier to output new observations, with or without being built-in. 

Right now a function like `pcv.fluor_fvfm` outputs a set of observations that are fixed, so running the same function again will overwrite the previous outputs. Example:

```python
outputs.add_observation(variable='fvfm_median', trait='Fv/Fm median',
                            method='plantcv.plantcv.fluor_fvfm', scale='none', datatype=float,
                            value=float(np.around(fvfm_median, decimals=4)), label='none')
```

If we modify the function(s) as you propose, with a parameter input that is used as a label, then the user could record multiple measurements. Example:

```python
outputs.add_observation(variable=parameter, trait=parameter,
                            method='plantcv.plantcv.photosynthesis.analyze_yii', scale='none', 
                            datatype=float,
                            value=float(np.around(yii_mean, decimals=4)), label='none')
```

The key thing is that `variable` needs to be unique (it's a dictionary key) in order to store it as a new observation. It only needs to be unique per entity though. An "entity" would generally be an image, but if you analyze multiple plants/objects per image then it could be an object/region also.

**Comment by dschneiderch** [comment]
Ok. I finally have a polished PSII pipeline that I think I am happy with (assuming I didn't overwrite my git branch somehow (!))
I will start a new branch for a new subpackage and we can iterate from there.

**Comment by nfahlgren** [comment]
Awesome, thanks @dschneiderch!

**Comment by dschneiderch** [comment]
first version of the new photosynthesis branch is up!
It works here and replicates what I was getting with my personal scripts.

- `analyze_npq` has arguments fmax, fm, mask, bins
- `analyze_yii` has arguments fdark, fmin, fmax, mask, bins, parameter
- i also created `qc_dark()` so fdark isn't qc'd anymore in analyze_yii. I left it as an argument because I wasn't sure if people use fdark to normalize fmin and fmax? 
- I left a couple place holder scripts for qL, qP, and fop. I wanted to make sure things were useable and yii and npq are much more common.
- I also created `masked_stats.py` in plantcv. I got tired of writing `np.mean(array[np.where(mask>0)])` so I defined `pcv.masked_stats.mean(array,mask)` (and std and median). This is only barely easier to write but more readable and easier to remember IMO. These functions are a little weird though because they are called with `pcv.masked_stats.mean()` but there is no folder called `masked_stats`. This might be unique to these functions and probably confusing for people. I just wanted to distinguish between `np.mean` and `masked_stats.mean()`. Is there a better way to do this?
- I'm going to setup a new project and test out the pipeline/Outputs that you have developed. I'm just working with a flat directory structure of images so it wasn't immediately obvious to me how one would refer to fm multiple times as you iterate through the images of an induction curve...
